### ISSUE TITLE
rank_hard_loss_layer.cpp

### ISSUE THREAD

**Issue by Robert0812** [xiaolonw/caffe-video_triplet#2]
How much amount of contribution of tloss2[https://github.com/xiaolonw/caffe-video_triplet/blob/master/src/caffe/layers/rank_hard_loss_layer.cpp#L171] in the forward process?

Dtype tloss2 = max(Dtype(0), dis_data[i * num + i + 1] - dis_data[(i + 1) * num + j] + Dtype(margin));

**Comment by kli-nlpr** [comment]
I have the same problem

**Comment by xiaolonw** [comment]
Just see it as an easy way to increase training samples.

**Comment by kli-nlpr** [comment]
@xiaolonw  that's make sense, thank you xiaolong

**Comment by icodingc** [comment]
@xiaolonw  
[rank_hard_loss.cpp](https://github.com/xiaolonw/caffe-video_triplet/blob/master/src/caffe/layers/rank_hard_loss_layer.cpp#L228 )第228行。
如果loss = max(0,||x - x1||**2 - ||x-x2||**2 + margin)的话，
我的理解是梯度应该这样计算：
dloss/dx = 2(x2-x1)
dloss/dx1 = -2(x-x1)
dloss/dx2 = 2(x-x2) 
但是代码中跟我理解的不一样，请问是我理解错么，还请解释一下。

**Comment by xiaolonw** [comment]
that is because we are using cosine distance, which means after normalization layer, ||x|| = 1, ||x1|| = 1. Thus loss = max(0, (2-2x*x1) - (2-2x*x2) + margin )

**Comment by icodingc** [comment]
thank you xiaolong.
